[motion-path] Incorrect anchoring when SVG renderers are used

nt1m opened the pull request:
#### 98a4d85150264a070b7957c27fd465eeb6a93366
<pre>
[motion-path] Incorrect anchoring when SVG renderers are used
<a href="https://bugs.webkit.org/show_bug.cgi?id=255684">https://bugs.webkit.org/show_bug.cgi?id=255684</a>
rdar://108285569

Reviewed by Darin Adler.

When we use SVG layout, and transform-box is not view-box, the current
implementation makes the object not align with offset-path because there
may be an offset (SVG x and y values) between the anchor point and offset-path.
We have to tweak the anchor point to fix the alignment.

This solution is analogous to Gecko&apos;s code: <a href="https://searchfox.org/mozilla-central/rev/47be6b9a0719f4311356695d9eed48de97b432b2/layout/base/MotionPathUtils.cpp#30">https://searchfox.org/mozilla-central/rev/47be6b9a0719f4311356695d9eed48de97b432b2/layout/base/MotionPathUtils.cpp#30</a>

We also split out TransformOperationData into its own files, to avoid circular dependencies.

This brings the Interop Motion Path score from 96% to 99%.

* LayoutTests/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::MotionPath::applyMotionPathTransform):
(WebCore::MotionPath::computePathForRay):
(WebCore::MotionPath::computePathForBox):
(WebCore::MotionPath::computePathForShape):
* Source/WebCore/rendering/MotionPath.h:
(WebCore::TransformOperationData::TransformOperationData): Deleted.
(WebCore::TransformOperationData::boundingBox const): Deleted.
(WebCore::TransformOperationData::motionPathData const): Deleted.
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/TransformOperationData.cpp: Added.
(WebCore::TransformOperationData::TransformOperationData):
* Source/WebCore/rendering/TransformOperationData.h: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyTransform const):
(WebCore::RenderStyle::applyCSSTransform const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/svg/SVGGraphicsElement.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/269642@main">https://commits.webkit.org/269642@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9bf3ca273488cfe40b106d1195dbc75daa4aef1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23150 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24269 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25068 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21405 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2686 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23688 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22257 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23390 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/811 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25917 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/641 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20957 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27123 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21080 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21221 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24991 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/652 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18423 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/583 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1061 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2941 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/859 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/TransformOperationData.h:
```diff
@@ -0,0 +1,56 @@
+
+class RenderElement;
+
+class TransformOperationData {
```
This feels to me like it should be a `struct`. Seems like overkill to have private data members and getter functions. I suppose though that we don’t need to support setting things independently. That could be expressed, though, by having it be `const` after construction instead of by using a class.

nt1m replied:
Changed it to a `struct`

darinadler commented on Source/WebCore/rendering/TransformOperationData.h:
```diff
@@ -0,0 +1,56 @@
+public:
+    TransformOperationData(FloatRect boundingBox, const RenderElement* renderer = nullptr);
+
+    TransformOperationData(FloatRect boundingBox, std::optional<MotionPathData> data, bool isSVGRenderer)
```
Would it be more efficient to move in the `MotionPathData` rather than copying it? Or is that a trivial structure?

nt1m replied:
MotionPathData is a small struct so it should be fine.

darinadler commented on Source/WebCore/rendering/TransformOperationData.h:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    const FloatRect& boundingBox() const { return m_boundingBox; }
+    const std::optional<MotionPathData> motionPathData() const { return m_motionPathData; }
```
The `const` here on the return type seems wrong. I think we either want a non-`const` type or a `const&`.

nt1m replied:
Removed after changing to a struct

nt1m committed:
[motion-path] Incorrect anchoring when SVG renderers are used
https://bugs.webkit.org/show_bug.cgi?id=255684
rdar://108285569

Reviewed by Darin Adler.

When we use SVG layout, and transform-box is not view-box, the current
implementation makes the object not align with offset-path because there
may be an offset (SVG x and y values) between the anchor point and offset-path.
We have to tweak the anchor point to fix the alignment.

This solution is analogous to Gecko's code: https://searchfox.org/mozilla-central/rev/47be6b9a0719f4311356695d9eed48de97b432b2/layout/base/MotionPathUtils.cpp#30

We also split out TransformOperationData into its own files, to avoid circular dependencies.

This brings the Interop Motion Path score from 96% to 99%.

* LayoutTests/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::MotionPath::applyMotionPathTransform):
(WebCore::MotionPath::computePathForRay):
(WebCore::MotionPath::computePathForBox):
(WebCore::MotionPath::computePathForShape):
* Source/WebCore/rendering/MotionPath.h:
(WebCore::TransformOperationData::TransformOperationData): Deleted.
(WebCore::TransformOperationData::boundingBox const): Deleted.
(WebCore::TransformOperationData::motionPathData const): Deleted.
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/TransformOperationData.cpp: Added.
(WebCore::TransformOperationData::TransformOperationData):
* Source/WebCore/rendering/TransformOperationData.h: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::applyTransform const):
(WebCore::RenderStyle::applyCSSTransform const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/svg/SVGGraphicsElement.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/269642@main

darinadler commented on Source/WebCore/rendering/TransformOperationData.h:
```diff
@@ -0,0 +1,49 @@
+class RenderElement;
+
+struct TransformOperationData {
+    TransformOperationData(FloatRect boundingBox, const RenderElement* renderer = nullptr);
```
We might want to go back here and add `explicit` so we don’t convert a rect into this accidentally.

nt1m replied:
https://github.com/WebKit/WebKit/pull/19416
